### PR TITLE
Integrate telemetry into `azd`

### PR DESCRIPTION
**Pull request by weikanglim**
- Telemetry system is now initialized at application startup in main
- `azd telemetry upload` is added as an internal/hidden command to upload telemetry in the background. A process is spun up once the user-interactive command terminates.
- Telemetry for each command invocation is now recorded

### CODE REVIEW

File: cli/azd/cmd/telemetry.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+			if telemetrySystem == nil {
+				return nil
+			}
+
+			return telemetrySystem.RunBackgroundUpload(cmd.Context(), rootOptions.EnableDebugLogging)
+		},
+	}
+	return cmd
+}
```
Comment (wbreza) [inline cli/azd/cmd/telemetry.go:43]: Any specific reason to create command manually here vs our Command builder?

File: cli/azd/internal/telemetry/telemetry.go [hunk 2]
```diff
@@ -18,13 +22,20 @@ import (
...
+
+// the equivalent of AZURE_CORE_COLLECT_TELEMETRY
+const collectTelemetryEnvVar = "AZURE_DEV_COLLECT_TELEMETRY"
+
 const telemetryItemExtension = ".trn"
-const devInstrumentationKey = "d3b9c006-3680-4300-9862-35fce9ac66c7"
-const prodInstrumentationKey = ""
+const (
+	devInstrumentationKey  = "d3b9c006-3680-4300-9862-35fce9ac66c7"
+	prodInstrumentationKey = ""
```
Comment (wbreza) [inline cli/azd/internal/telemetry/telemetry.go:36]: Anything to worry about that we are exposing these keys directly in the source?

File: cli/azd/cmd/telemetry.go (hunk 1 above)
Comment (wbreza) [reply cli/azd/cmd/telemetry.go:43]: Avoiding to use the builder that will eventually be used to instrument the other commands thus avoiding a circular loop?

File: cli/azd/internal/telemetry/tracer.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+// and timed operation of a workflow that is traced. A Tracer is used to
+// create a Span and it is then up to the operation the Span represents to
+// properly end the Span when the operation itself ends.
+type Span interface {
+	redefinedSpan
+}
+
+type wrapperSpan struct {
+	span trace.Span
+}
```
Comment (wbreza) [inline cli/azd/internal/telemetry/tracer.go:88]: What is the use case for the `wrapperSpan` vs the standard `Span` ?

File: cli/azd/internal/telemetry/tracer.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	TracerProvider() trace.TracerProvider
+}
+
+// Span is the individual component of a trace. It represents a single named
+// and timed operation of a workflow that is traced. A Tracer is used to
+// create a Span and it is then up to the operation the Span represents to
+// properly end the Span when the operation itself ends.
+type Span interface {
+	redefinedSpan
+}
```
Comment (wbreza) [inline cli/azd/internal/telemetry/tracer.go:84]: Why reference the internal interface here vs just exposing the public `Span` interface with the methods directly?

File: cli/azd/internal/telemetry/tracer.go (hunk 3 above)
Comment (wbreza) [reply cli/azd/internal/telemetry/tracer.go:88]: OK - so this is just wrapping the otel `Span` ?

**Review by wbreza — APPROVED** [review]
Looks great overall! Just have a few questions.  Nothing blocking.

File: cli/azd/internal/telemetry/tracer.go (hunk 4 above)
Comment (weikanglim) [reply cli/azd/internal/telemetry/tracer.go:84]: The idea here is to create separation between the base `tracer.Span` and any functionality we're extending for our own use case. Any new functionality would be added here in `Span`. And once we support the functionality of `Span` fully, we would delete `redefinedSpan`. Basically, keeping the forked vs. extended code clear

File: cli/azd/internal/telemetry/tracer.go (hunk 3 above)
Comment (weikanglim) [reply cli/azd/internal/telemetry/tracer.go:88]: Yep, for any extended functionality

File: cli/azd/internal/telemetry/telemetry.go (hunk 2 above)
Comment (weikanglim) [reply cli/azd/internal/telemetry/telemetry.go:36]: No, it's the unfortunate reality of client-side telemetry and is pretty much standard implementation. You have to expose some credentials to the client. The only authorized operation is to push new logs which is fine

**Pull request merged**